client/state/timezones/test/reducer.js

```diff
@@ -0,0 +1,335 @@
+/**
+ * External dependencies
+ */
+import { expect } from 'chai';
+import deepFreeze from 'deep-freeze';
+
+/**
+ * Internal dependencies
+ */
+import { useSandbox } from 'test/helpers/use-sinon';
+
+import timezonesReducer, {
+	byContinentsReducer,
+	labelsReducer,
+	rawOffsetReducer,
+} from '../reducer';
+
+import { timezonesReceive } from '../actions';
+import { fromApi } from 'state/data-layer/wpcom/timezones';
+
+/**
+ * Fixture data
+ */
+import {
+	RAW_OFFSETS,
+	TIMEZONES,
+	TIMEZONES_BY_CONTINENT,
+} from './fixture';
+
+describe( 'reducer', () => {
+	let sandbox;
+
+	useSandbox( newSandbox => {
+		sandbox = newSandbox;
+		sandbox.stub( console, 'warn' );
+	} );
+
+	it( 'should export expected reducer keys', () => {
+		expect( timezonesReducer( undefined, {} ) ).to.have.keys( [
+			'byContinents',
+			'labels',
+			'rawOffsets',
+		] );
+	} );
+
+	describe( '#rawOffsetReducer()', () => {
+		it( 'should default to an empty Object', () => {
+			expect( rawOffsetReducer( undefined, {} ) ).to.eql( {} );
+		} );
+
+		it( 'should index timezones.rawOffsets state', () => {
+			const initialState = undefined;
+			const restAPIResponse = {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			};
+
+			const apiResponse = fromApi( restAPIResponse );
+			const action = timezonesReceive( apiResponse );
+
+			const expectedState = apiResponse.rawOffsets;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
+
+		it( 'should override timezones.rawOffsets state', () => {
+			const initialState = fromApi( {
+				manual_utc_offsets: [
+					{ value: 'foo', label: 'bar' }
+				],
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			} ).rawOffsets;
+
+			deepFreeze( initialState );
+
+			const restAPIResponse = {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			};
+
+			const apiResponse = fromApi( restAPIResponse );  // data-layer processing
+			const action = timezonesReceive( apiResponse );
+
+			const expectedState = apiResponse.rawOffsets;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
+
+		it( 'should persist state', () => {
+			const initialState = fromApi( {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			} );
+			deepFreeze( initialState );
+
+			const action = { type: 'SERIALIZE' };
+
+			const expectedState = initialState;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
```

Comment:
> But the second part is, how would you verify that the JSON schema that the reducer uses actually matches the expected state shape?

probably belongs in the reducer area. it's also in the gray zone where testing is murkier. if the schema itself defines the structure then we don't need to test it because JSON schema has its own testing and validation.

so how do we make sure that the reducer matches the schema? we add a test duplicating the structure of the schema and the reducers. how do we make sure that the mocks in the tests match the schema? we add a test to test the test's mock data? this is where I get reluctant because we veer away from what unit tests are designed to do. a hard-coded mock in a test isn't any more resilient than a hard-coded structure in a reducer.

we don't have an integration suite to my knowledge. maybe the end-to-end testing (is that you @alisterscott?) is something I should learn more about. Iv'e heard many people talk about running slower integration tests in a separate process from development. unit tests are fast, isolated, and support development; integration tests run periodically independent from code changes and catch unintentional bugs like these